File: internal/models/activity_log.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
)

// ActivityLog represents a logged activity for a pet
type ActivityLog struct {
	ID           uuid.UUID      `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	PetID        uuid.UUID      `gorm:"type:uuid;not null;index" json:"pet_id"`
	PetDiseaseID *uuid.UUID     `gorm:"type:uuid;index" json:"pet_disease_id,omitempty"`
	ActivityType string         `gorm:"column:activity_type;not null;size:100;index" json:"activity_type"`
	LoggedAt     time.Time      `gorm:"column:logged_at;not null;index" json:"logged_at"`
	ActivityData datatypes.JSON `gorm:"column:activity_data;type:jsonb;not null;default:'{}'" json:"activity_data"`
	Notes        string         `gorm:"type:text" json:"notes,omitempty"`
	LoggedBy     *uuid.UUID     `gorm:"type:uuid;index" json:"logged_by,omitempty"`
	CreatedAt    time.Time      `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt    time.Time      `gorm:"autoUpdateTime" json:"updated_at"`

	// Relations
	Pet        Pet             `gorm:"foreignKey:PetID" json:"pet,omitempty"`
	PetDisease *PetDisease     `gorm:"foreignKey:PetDiseaseID" json:"pet_disease,omitempty"`
	Logger     *User           `gorm:"foreignKey:LoggedBy" json:"logger,omitempty"`
	Photos     []ActivityPhoto `gorm:"foreignKey:ActivityLogID" json:"photos,omitempty"`
}

func (ActivityLog) TableName() string {
	return "activity_logs"
}

// ActivityPhoto stores photos attached to activity logs
type ActivityPhoto struct {
	ID            uuid.UUID `gorm:"type:uuid;primary_key;default:gen_random_uuid()" json:"id"`
	ActivityLogID uuid.UUID `gorm:"type:uuid;not null;index" json:"activity_log_id"`
	PhotoURL      string    `gorm:"column:photo_url;type:text;not null" json:"photo_url"`
	ThumbnailURL  string    `gorm:"column:thumbnail_url;type:text" json:"thumbnail_url,omitempty"`
	Caption       string    `gorm:"type:text" json:"caption,omitempty"`
	FileSize      int       `gorm:"column:file_size" json:"file_size,omitempty"`
	CreatedAt     time.Time `gorm:"autoCreateTime" json:"created_at"`

	// Relations
	ActivityLog ActivityLog `gorm:"foreignKey:ActivityLogID" json:"activity_log,omitempty"`
}

func (ActivityPhoto) TableName() string {
	return "activity_photos"
}

// Common activity types
const (
	ActivityTypeWaterIntake  = "water_intake"
	ActivityTypeFoodIntake   = "food_intake"
	ActivityTypeWeight       = "weight"
	ActivityTypeMedication   = "medication"
	ActivityTypeUrination    = "urination"
	ActivityTypeDefecation   = "defecation"
	ActivityTypeVomiting     = "vomiting"
	ActivityTypeExercise     = "exercise"
	ActivityTypeSleep        = "sleep"
	ActivityTypeSymptom      = "symptom"
	ActivityTypeBehavior     = "behavior"
	ActivityTypeVitalSigns   = "vital_signs"
	ActivityTypeBloodGlucose = "blood_glucose"
	ActivityTypeAppetite     = "appetite"
	ActivityTypeMood         = "mood"
)

File: mobile_app/lib/data/models/activity_log.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'activity_log.freezed.dart';
part 'activity_log.g.dart';

/// Common activity type constants
class ActivityTypes {
  static const String waterIntake = 'water_intake';
  static const String foodIntake = 'food_intake';
  static const String weight = 'weight';
  static const String medication = 'medication';
  static const String urination = 'urination';
  static const String defecation = 'defecation';
  static const String vomiting = 'vomiting';
  static const String exercise = 'exercise';
  static const String sleep = 'sleep';
  static const String symptom = 'symptom';
  static const String behavior = 'behavior';
  static const String vitalSigns = 'vital_signs';
  static const String bloodGlucose = 'blood_glucose';
  static const String appetite = 'appetite';
  static const String mood = 'mood';
}

@freezed
class ActivityLog with _$ActivityLog {
  const ActivityLog._();

  const factory ActivityLog({
    required String id,
    @JsonKey(name: 'pet_id') required String petId,
    @JsonKey(name: 'pet_disease_id') String? petDiseaseId,
    @JsonKey(name: 'activity_type') required String activityType,
    @JsonKey(name: 'logged_at') required DateTime loggedAt,
    @JsonKey(name: 'activity_data') @Default({}) Map<String, dynamic> activityData,
    String? notes,
    @JsonKey(name: 'logged_by') String? loggedBy,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
    List<ActivityPhoto>? photos,
  }) = _ActivityLog;

  factory ActivityLog.fromJson(Map<String, dynamic> json) => _$ActivityLogFromJson(json);

  /// Get numeric value from activity_data
  double? get value {
    final val = activityData['value'];
    if (val == null) return null;
    if (val is num) return val.toDouble();
    return null;
  }

  /// Get unit from activity_data
  String? get unit => activityData['unit'] as String?;

  /// Get input type from activity_data
  String? get inputType => activityData['input_type'] as String?;

  /// Check if this is a checkbox type activity
  bool get isChecked => activityData['checked'] == true;
}

@freezed
class ActivityPhoto with _$ActivityPhoto {
  const factory ActivityPhoto({
    required String id,
    @JsonKey(name: 'activity_log_id') required String activityLogId,
    @JsonKey(name: 'photo_url') required String photoUrl,
    @JsonKey(name: 'thumbnail_url') String? thumbnailUrl,
    String? caption,
    @JsonKey(name: 'file_size') int? fileSize,
    @JsonKey(name: 'created_at') DateTime? createdAt,
  }) = _ActivityPhoto;

  factory ActivityPhoto.fromJson(Map<String, dynamic> json) => _$ActivityPhotoFromJson(json);
}
